ndcube/meta.py

```diff
@@ -0,0 +1,466 @@
+        idx = [not isinstance(i, numbers.Integral) and s == 0 for i, s in zip(item, data_shape)]
+        idx = np.arange(len(idx))[idx]
+        item[idx] = np.array([slice(None)] * len(idx))
+
+        # Edit data shape and calculate which axis will be dropped.
+        dropped_axes = np.zeros(naxes, dtype=bool)
+        new_shape = new_meta.data_shape
+        for i, axis_item in enumerate(item):
+            if isinstance(axis_item, numbers.Integral):
+                dropped_axes[i] = True
+            elif isinstance(axis_item, slice):
+                start = axis_item.start
+                if start is None:
+                    start = 0
+                if start < 0:
+                    start = data_shape[i] - start
+                stop = axis_item.stop
+                if stop is None:
+                    stop = data_shape[i]
+                if stop < 0:
+                    stop = data_shape[i] - stop
+                new_shape[i] = stop - start
+            else:
+                raise TypeError("Unrecognized slice type. "
+                                "Must be an int, slice and tuple of the same.")
+        kept_axes = np.invert(dropped_axes)
+        new_meta._data_shape = new_shape[kept_axes]
+
+        # Slice all metadata associated with axes.
+        for key, value in self.meta.items():
+            axis = self.meta.axes.get(key, None)
+            drop_key = False
+            if axis is None:
+                continue
+            # Calculate new axis indices.
+            new_axis = np.asarray(list(
+                set(axis).intersection(set(np.arange(naxes)[kept_axes]))
+                ))
+            if len(new_axis) == 0:
+                new_axis = None
+            else:
+                cumul_dropped_axes = np.cumsum(dropped_axes)[new_axis]
+                new_axis -= cumul_dropped_axes
+
+            # Calculate sliced metadata values.
+            axis_shape = tuple(self.meta.data_shape[axis])
+            if _is_scalar(value):
+                new_value = value
+                # If scalar metadata's axes have been dropped, mark metadata to be dropped.
+                if new_axis is None:
+                    drop_key = True
+            else:
+                value_is_axis_aligned = _is_axis_aligned(value, axis_shape)
+                if value_is_axis_aligned:
+                    new_item = kept_axes[axis]
+                else:
+                    new_item = tuple(item[axis])
+                # Slice metadata value.
+                try:
+                    new_value = value[new_item]
+                except:
+                    # If value cannot be sliced by fancy slicing, convert it
+                    # it to an array, slice it, and then if necessary, convert
+                    # it back to its original type.
+                    new_value = (np.asanyarray(value)[new_item])
+                    if hasattr(new_value, "__len__"):
+                        new_value = type(value)(new_value)
```

Comment:
Did we say we were going to nuke this bit?

Reply:
I left a comment about this somewhere after our talk.  I realised that without this, `NDMeta` could not support dropping axes as part of slicing if it contained lists or tuples as axis-aligned metadata.  So as ugly as it is, I think it's better to keep it.